Don't assume functions starting with _ are ES5 constructors

**What version of ESLint are you using?**

2.3.0 or later (any version including 13c7c19f).

**What parser (default, Babel-ESLint, etc.) are you using?**

Default

**Please show your full configuration:**

```yml
rules:
  consistent-return: 2
```

**What did you do? Please include the actual source code causing the issue.**

```js
function _doSomething(condition) {
    if (condition) {
        return true;
    }
}
```

**What did you expect to happen?**

```
  1:10  error  Expected to return a value at the end of this function  consistent-return
```

**What actually happened? Please include the actual, raw output from ESLint.**

No output is produced.

This is an unintended result of my change in #5398 which excludes functions which pass `isES5Constructor`.  The logic in `isES5Constructor` checks that [the first character of the function name is invariant under `.toLocaleUpperCase`](https://github.com/eslint/eslint/blob/v2.9.0/lib/ast-utils.js#L60), which includes both upper-case and caseless characters (including punctuation).  This is especially unfortunate since `_` is commonly used to indicate semi-private methods.

I suggest excluding `_` (and maybe `$`) from `isES5Constructor`.  In the cases where a constructor begins with these characters, the user can add inline config to disable `consistent-return` and `no-invalid-this` as needed.  If that sounds agreeable, I'll work up a PR.

Thoughts?

CC: @ilyavolodin, @mysticatea, @nzakas since you guys helped with #5398.

--- a/lib/ast-utils.js
+++ b/lib/ast-utils.js
@@ -57,7 +57,7 @@ function isModifyingReference(reference, index, references) {
 function isES5Constructor(node) {
     return (
         node.id &&
-        node.id.name[0] === node.id.name[0].toLocaleUpperCase()
+        node.id.name[0] !== node.id.name[0].toLocaleLowerCase()
     );
 }
 

--- a/tests/lib/rules/consistent-return.js
+++ b/tests/lib/rules/consistent-return.js
@@ -113,6 +113,16 @@ ruleTester.run("consistent-return", rule, {
             ]
         },
         {
+            code: "function _foo() { if (a) return true; }",
+            errors: [
+                {
+                    message: "Expected to return a value at the end of this function.",
+                    type: "FunctionDeclaration",
+                    column: 10
+                }
+            ]
+        },
+        {
             code: "f(function foo() { if (a) return true; });",
             errors: [
                 {

